lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp

```diff
@@ -535,19 +535,54 @@ class WhenOpVisitor : public LastConnectResolver<WhenOpVisitor> {
   /// scope, i.e. not in a WhenOp region, then there is no condition.
   Value andWithCondition(Operation *op, Value value) {
     // 'and' the value with the current condition.
-    return OpBuilder(op).createOrFold<AndPrimOp>(
+    OpBuilder builder(op);
+    return andWithCondition(builder, value);
+  }
+
+  /// And a 1-bit value with the current condition.  If we are in the outer
+  /// scope, i.e. not in a WhenOp region, then there is no condition.
+  Value andWithCondition(OpBuilder &builder, Value value) {
+    // 'and' the value with the current condition.
+    return builder.createOrFold<AndPrimOp>(
+        condition.getLoc(), condition.getType(), condition, value);
+  }
+
+  // Or a 1-bit value with the current condition.  If we are in the outer
+  /// scope, i.e. not in a WhenOp region, then there is no condition.
+  Value orWithCondition(OpBuilder &builder, Value value) {
+    // 'or' the value with the current condition.
+    return builder.createOrFold<OrPrimOp>(
         condition.getLoc(), condition.getType(), condition, value);
   }
 
-  // Create an implication from the condition to the given value
-  // This is implemented as (or (not condition) value) to avoid
-  // the corner case where value is an implication.
-  Value impliesLTLCondition(Operation *op, Value value) {
-    OpBuilder builder = OpBuilder(op);
-    Value notCond = builder.createOrFold<LTLNotIntrinsicOp>(
-        condition.getLoc(), condition.getType(), condition);
-    return builder.createOrFold<LTLOrIntrinsicOp>(
-        condition.getLoc(), condition.getType(), notCond, value);
+  template <typename Op>
+  void visitVerifIntrinsic(Op op) {
+    OpBuilder builder(op);
+    Value foldedWhen;
+    LTLClockIntrinsicOp clockOp;
+    LTLDisableIntrinsicOp disable;
+    // Extract the clock and disable from the UInt<1>
+    if ((clockOp =
+             dyn_cast<LTLClockIntrinsicOp>(op.getProperty().getDefiningOp()))) {
+      if ((disable = dyn_cast<LTLDisableIntrinsicOp>(
+               clockOp.getInput().getDefiningOp()))) {
+
+        Value newDisable = orWithCondition(builder, disable.getDisable());
+        foldedWhen = builder.create<LTLDisableIntrinsicOp>(
+            op.getLoc(), newDisable.getType(), disable.getInput(), newDisable);
+      }
+      foldedWhen = builder.create<LTLClockIntrinsicOp>(
+          op.getLoc(), foldedWhen.getType(), foldedWhen, clockOp.getClock());
+    }
+    if (foldedWhen) {
+      op.getPropertyMutable().assign(foldedWhen);
+
+      // destroy the old clock and disable ops
+      if (clockOp && isOpTriviallyDead(clockOp))
+        clockOp.erase();
+      if (disable && isOpTriviallyDead(disable))
+        disable.erase();
+    }
```

Comment:
I'm worried about the pattern matching affecting semantics of when, e.g. inserting a node for clock intrinsic changes a semantics of two apparently equivalent IR. 

```mlir
firrtl.circuit "verifassert" {
firrtl.module @verifassert(
  in %clock : !firrtl.clock, in %cond : !firrtl.uint<1>, in %prop : !firrtl.uint<1>, in %prop1 : !firrtl.uint<1>,
  in %disable : !firrtl.uint<1>, in %disable1 : !firrtl.uint<1>, in %reset : !firrtl.uint<1>
) {
  firrtl.when %cond : !firrtl.uint<1> {

    %dis_prop = firrtl.int.ltl.disable %prop, %disable : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<1>
    %clk_prop = firrtl.int.ltl.clock %dis_prop, %clock : (!firrtl.uint<1>, !firrtl.clock) -> !firrtl.uint<1>
    %clk_prop_node = firrtl.node %clk_prop: !firrtl.uint<1>
    firrtl.int.verif.assert %clk_prop_node : !firrtl.uint<1>
    firrtl.int.verif.assert %clk_prop : !firrtl.uint<1>
  } else {
  }
}
}
```

```verilog
  assert property (@(posedge clock) disable iff (disable_0) prop);      // bar.mlir:9:17, :10:17, :12:5
  wire _GEN = cond | disable_0; // bar.mlir:4:33
  assert property (@(posedge clock) disable iff (_GEN) prop);   // bar.mlir:4:33, :13:5
```

We could look through nodes/wires but I guess it's still fragile (for example running  canonicalization before ExpandWhen might change the semantics). I agree that intrinsic/foreign type is actually good solution here.

Reply:
I'm not entirely sure I see what the issue is here, could you elaborate @uenoku ?

Reply:
In the first assert ` firrtl.int.verif.assert %clk_prop_node` a condition (`%cond`) is dropped in ExpandWhen.     Whereas for the second assert `firrtl.int.verif.assert %clk_prop`, `%cond` is preserved.  I think it's problematic that existence of node affects the semantics.